Add support for deploying Agent on k8s with kustomize (#2754)
elastic/elastic-agent: deploy/kubernetes/README.md

This is the list of officially supported Elastic Agents, with example manifests to run
them in Kubernetes:

Agent Mode | Description
---- | ----
[Elastic Agent managed](elastic-agent-managed) | Elastic Agent managed by Fleet setup
[Elastic Agent standalone](elastic-agent-standalone) | Standalone Elastic Agent setup

> > Note: The kube-state-metrics (KSM) simple service is not installed as part of the above manifests and needs to be installed from [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics).

Reviewer: ```suggestion
> Note: `kube-state-metrics` (KSM) is not installed as part of the above manifests and needs to be installed from [kube-state-metrics](https://github.com/kubernetes/kube-state-metrics).
```

## Kustomize Templates

In addition to the above manifests, the list below includes the official [kustomize](https://github.com/kubernetes-sigs/kustomize) templates to run them in kubernetes:

Agent Scenario | Description
---- | ----
[Elastic Agent managed - Default ](./default/elastic-agent-managed/) | Default Elastic Agent managed by Fleet setup. Includes the installation of kube-state-metrics (KSM) in default configuration.
[Elastic Agent standalone Default ](./default/elastic-agent-standalone/) | Default Standalone Elastic Agent setup
[Elastic Agent managed - With KSM in autosharding configuration ](./ksm-autosharding/elastic-agent-managed/) | Elastic Agent managed by Fleet setup with [KSM in autosharding configuration](https://github.com/kubernetes/kube-state-metrics#automated-sharding)
[Elastic Agent standalone - With KSM in autosharding configuration](./ksm-autosharding/elastic-agent-standalone/) | Standalone Elastic Agent setup with [KSM in autosharding configuration](https://github.com/kubernetes/kube-state-metrics#automated-sharding)

> 

Reviewer: We need to mention here briefly what each scenario is for or just link to the section below where autosharding mode is explained.

Author: Updated, how about now?

### How to use the kustomize templates

> Navigate to scenario you need to install: 

Reviewer: ```suggestion
Navigate to the scenario you want to install: 
```

> E.g. we need to install the default Elastic Agent Scenario with kustomize.

Reviewer: ```suggestion
E.g. we want to install the default Elastic Agent Scenario with kustomize.
```

```bash
cd elastic-agent-kustomize/default/elastic-agent-standalone
kubectl apply -k .
```

> Since Kubernetes version 1.14, Kubectl supports the use of kustomize

### Kube State Metrics (KSM) in autosharding configuration

The several challenges phased during scaling of Kubernetes cluster are [explained here](https://github.com/elastic/ingest-docs/blob/main/docs/en/ingest-management/elastic-agent/scaling-on-kubernetes.asciidoc)

> The suggested method of installing Elastic Agent in large Kubernetes clusters is to install Elastic Agent as a Side Container with KSM in order to collect the KSM metrics. As Kubernetes cluster becomes larger, we increase the number of KSM shards and we have a dedicated Elastic Agent to collect from specific shard endpoint. Those are installed as part of [Statefulset](./elastic-agent-kustomize/ksm-autosharding/elastic-agent-standalone/base/elastic-agent-standalone-ksm-statefulset-configmap.yaml)

Reviewer: ```suggestion
The suggested method of running Elastic Agent in large Kubernetes clusters is to install Elastic Agent as a sidecar container with KSM in order to collect the KSM metrics. As a Kubernetes cluster becomes larger, we need to increase the number of KSM shards and we need to have a dedicated Elastic Agent in order to collect from the specific shard's endpoint. Those are installed as part of a [Statefulset](./elastic-agent-kustomize/ksm-autosharding/elastic-agent-standalone/base/elastic-agent-standalone-ksm-statefulset-configmap.yaml)
```

Also can we elaborate more on this Statefulset approach?

> In addition to the Elastic Agents for KSM, the rest of the metrics are collected from Elastic Agents deployed as [daemonsets](./elastic-agent-kustomize/ksm-autosharding/elastic-agent-standalone/base/elastic-agent-standalone-ksm-daemonset-configmap.yaml)

Reviewer: ```suggestion
Apart from the KSM metrics, the rest of the metrics are collected from Elastic Agents deployed as [daemonsets](./elastic-agent-kustomize/ksm-autosharding/elastic-agent-standalone/base/elastic-agent-standalone-ksm-daemonset-configmap.yaml)
```

> More information for the configuration of Elastic Agent can be found [here](https://github.com/elastic/elastic-agent/blob/main/docs/elastic-agent-ksm-sharding.md)

Reviewer: Why `main` and not the related version? If I look into this current file from 8.9 then clicking this link will take me to `main`, right? This is inconsistent. 

```suggestion
More information about running Elastic Agent along with KSM in sharding mode see [here](https://github.com/elastic/elastic-agent/blob/main/docs/elastic-agent-ksm-sharding.md)
```

This is the list of officially supported Elastic Agents, with example manifests to run
them in Kubernetes:

Agent Mode | Description
---- | ----
[Elastic Agent managed](elastic-agent-managed) | Elastic Agent managed by Fleet setup
[Elastic Agent standalone](elastic-agent-standalone) | Standalone Elastic Agent setup

> > Note: Kube State Metrics Library is not installed as part of the above manifests and needs to be installed additionally

Reviewer 2: ```suggestion
> Note: The kube-state-metrics (KSM) simple service is not installed as part of the above manifests and needs to be installed from https://github.com/kubernetes/kube-state-metrics.
```

## Kustomize Templates

> Additional to the above manifests, the list below includes the official [kustomize](https://github.com/kubernetes-sigs/kustomize) templates to run them in kubernetes:

Reviewer 2: ```suggestion
In addition to the above manifests, the list below includes the official [kustomize](https://github.com/kubernetes-sigs/kustomize) templates to run them in kubernetes:
```

*(KSM = Kube State Metrics)

### How to use the kustomize templates

Navigate to scenario you need to install: 

E.g. we need to install the default Elastic Agent Scenario with kustomize.

```bash
cd elastic-agent-kustomize/default/elastic-agent-standalone
kubectl apply -k .
```

> Since Kubernetes version 1.14, Kubectl supports the use of kustomize

### Kube State Metrics (KSM) in autosharding configuration

The several challenges phased during scaling of Kubernetes cluster are [explained here](https://github.com/elastic/ingest-docs/blob/main/docs/en/ingest-management/elastic-agent/scaling-on-kubernetes.asciidoc)

The suggested method of installing Elastic Agent in large Kubernetes clusters is to install Elastic Agent as a Side Container with KSM in order to collect the KSM metrics. As Kubernetes cluster becomes larger, we increase the number of KSM shards and we have a dedicated Elastic Agent to collect from specific shard endpoint. Those are installed as part of [Statefulset](./elastic-agent-kustomize/ksm-autosharding/elastic-agent-standalone/base/elastic-agent-standalone-ksm-statefulset-configmap.yaml)

> Additional to the Elastic Agents for KSM, the rest of metrics are collected from Elastic Agents deployed as [daemonsets](./elastic-agent-kustomize/ksm-autosharding/elastic-agent-standalone/base/elastic-agent-standalone-ksm-daemonset-configmap.yaml)

Reviewer 2: ```suggestion
In addition to the Elastic Agents for KSM, the rest of the metrics are collected from Elastic Agents deployed as [daemonsets](./elastic-agent-kustomize/ksm-autosharding/elastic-agent-standalone/base/elastic-agent-standalone-ksm-daemonset-configmap.yaml)
```

## Kustomize Templates

In addition to the above manifests, the list below includes the official [kustomize](https://github.com/kubernetes-sigs/kustomize) templates to run them in Kubernetes:

Agent Scenario | Description
---- | ----
[Elastic Agent managed - Default ](./default/elastic-agent-managed/) | Default Elastic Agent managed by Fleet setup. Kube-state-metrics (KSM) is installed automatically.
[Elastic Agent standalone Default ](./default/elastic-agent-standalone/) | Default Standalone Elastic Agent setup. Kube-state-metrics (KSM) is installed automatically.
[Elastic Agent managed - With KSM in autosharding configuration ](./ksm-autosharding/elastic-agent-managed/) | Elastic Agent managed by Fleet setup with KMS in [autosharding configuration](https://github.com/kubernetes/kube-state-metrics#automated-sharding)
[Elastic Agent standalone - With KSM in autosharding configuration](./ksm-autosharding/elastic-agent-standalone/) | Standalone Elastic Agent setup with KSM in [autosharding configuration](https://github.com/kubernetes/kube-state-metrics#automated-sharding)

How to choose the appropriate scenario:

> - `Default` Elastic agent scenario, deploys Elastic Agent as Daemonset. The default installation method, that will deploy one agent per Kubernetes node.

Reviewer: ```suggestion
- `Default` scenario, deploys Elastic Agent as Daemonset. The default installation method, that will deploy one agent per Kubernetes node.
```

### Kube State Metrics (KSM) in autosharding configuration

The suggested method of running Elastic Agent in large Kubernetes clusters is to install Elastic Agent as a sidecar container with KSM in order to collect the KSM metrics. As a Kubernetes cluster becomes larger, we need to increase the number of KSM shards and also need to have a dedicated Elastic Agent in order to collect from the specific shard's endpoint. Those are installed as part of a [Statefulset](./elastic-agent-kustomize/ksm-autosharding/elastic-agent-standalone/base/elastic-agent-standalone-ksm-statefulset-configmap.yaml). In more details, the Statefulset acts as a patch of the current KSM Statefuleset pod, meaning that it is responsible to install Elastic Agent with kube-state-metrics datastream configuration as a second container within the same pod. It includes `hostNetwork:false`. The KSM and Elastic Agent will share the same localhost network to communicate.

Apart from the KSM metrics, the rest of the metrics are collected from Elastic Agents deployed as [daemonsets](./elastic-agent-kustomize/ksm-autosharding/elastic-agent-standalone/base/elastic-agent-standalone-ksm-daemonset-configmap.yaml)

> More information about running Elastic Agent along with KSM in sharding mode see in [elastic-agent-ksm-sharding](https://github.com/elastic/elastic-agent/blob/main/docs/elastic-agent-ksm-sharding.md)

Reviewer: ```suggestion
More information about running Elastic Agent along with KSM in sharding mode can be found at [elastic-agent-ksm-sharding](https://github.com/elastic/elastic-agent/blob/main/docs/elastic-agent-ksm-sharding.md)
```
